src/datachain/lib/dc/storage.py

```diff
@@ -79,40 +102,73 @@ def from_storage(
     cache = session.catalog.cache
     client_config = session.catalog.client_config
 
-    list_ds_name, list_uri, list_path, list_ds_exists = get_listing(
-        uri, session, update=update
-    )
+    uris = uri if isinstance(uri, (tuple, list)) else [uri]
+
+    if not uris:
+        raise ValueError("No URIs provided")
 
-    # ds_name is None if object is a file, we don't want to use cache
-    # or do listing in that case - just read that single object
-    if not list_ds_name:
-        dc = from_values(
+    storage_chain = None
+    listed_uris = set()
+    file_values = []
+
+    for single_uri in uris:
+        update_for_uri = update
+        if update:
+            # If update is True, we need to check if the URI has already been listed
+            # If it has, we don't need to update the listing
+            _, lst_uri, _ = parse_listing_uri(os.fsdecode(single_uri), client_config)
+            if lst_uri in listed_uris:
+                listed_uris.add(lst_uri)
+                update_for_uri = False
+
+        list_ds_name, list_uri, list_path, list_ds_exists = get_listing(
+            single_uri, session, update=update_for_uri
+        )
+
+        # list_ds_name is None if object is a file, we don't want to use cache
+        # or do listing in that case - just read that single object
+        if not list_ds_name:
+            file_values.append(
+                get_file_info(list_uri, cache, client_config=client_config)
+            )
+            continue
+
+        if update_for_uri or not list_ds_exists:
+            # disable prefetch for listing, as it pre-downloads all files
+            (
+                from_records(
+                    DataChain.DEFAULT_FILE_RECORD,
+                    session=session,
+                    settings=settings,
+                    in_memory=in_memory,
+                )
+                .settings(prefetch=0)
+                .gen(
```

Comment:
May be naive question, but still. Is it possible to use single generator to generate listings for all the URIs here? 🤔